### PR TITLE
Bugfix: find name uses within classes

### PR DESCRIPTION
**Pull request by joyceerhl**
For #23 

With this PR, the fixed notebook looks like this:
bugfix

@andrewhead feedback is much appreciated! :-) And thank you for your patience and help so far!

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
     "moduleResolution": "node",
     "target": "es6",
     "outDir": "./lib",
-    "lib": ["es6", "dom"],
+    "lib": ["es6", "dom", "esnext.array"],
```
Comment (andrewhead) [inline tsconfig.json]: What does this extra lib in tsconfig.json get us?

File: src/analysis/slice/data-flow.ts [hunk 2]
```diff
@@ -337,6 +337,11 @@ export class DataflowAnalyzer {
         uses = undefinedRefs.filter(r => r.level == ReferenceType.USE);
         break;
       case ast.CLASS:
+        // For each class function, call getUses
```
Comment (andrewhead) [inline src/analysis/slice/data-flow.ts]: nit-pick on comment: won't this call getUses for every statement in the function (including class variable defs?)

File: src/analysis/slice/data-flow.ts [hunk 3]
```diff
@@ -337,6 +337,11 @@ export class DataflowAnalyzer {
...
+        const usesArr = statement.code
+          .flatMap(classStatement => this.getUses(classStatement, _));
```
Comment (andrewhead) [inline src/analysis/slice/data-flow.ts]: Great. Simple. Effective. Nice!

File: tsconfig.json (hunk 1 above)
Comment (joyceerhl) [reply tsconfig.json]: According to [this thread](https://stackoverflow.com/questions/53556409/typescript-flatmap-flat-flatten-doesnt-exist-on-type-any), the [`flatMap` function](https://github.com/Microsoft/TypeScript/blob/ef7997cb73ce46081cbfebc0edd2a0295ba30c33/lib/lib.esnext.array.d.ts).

Without this, TypeScript complains that `Property 'flatMap' does not exist on type 'ISyntaxNode[]'`.

File: src/analysis/slice/data-flow.ts (hunk 2 above)
Comment (joyceerhl) [reply src/analysis/slice/data-flow.ts]: Updated the comment!

File: tsconfig.json (hunk 1 above)
Comment (andrewhead) [reply tsconfig.json]: Got it! Possibly stupid question: does this mean this code won't work for browsers that don't support `flatMap` on array (e.g., Chrome pre-69, Firefox pre-62 according to [this MDN page](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap))? Or will `flatMap` get cross-browser compatible JavaScript by the TypeScript compiler?

If this breaks compatibility with older browsers, we probably want to consider an alternate implementation that doesn't use the flatMap method.

Comment (joyceerhl) [reply tsconfig.json]: [You're right](https://stackoverflow.com/questions/39837678/why-no-array-prototype-flatmap-in-javascript#comment99750916_56586455), I forgot this needs to run in a browser. I don't know much about browser compatibility, would [this](https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_flattendeep) implementation be sufficiently compatible?

File: src/analysis/slice/data-flow.ts (hunk 3 above)
Comment (andrewhead) [inline src/analysis/slice/data-flow.ts]: A potential fix is to just factor out `flatMap`. How about...

```suggestion
          .forEach(classStatement => uses.add(...this.getUses(classStatement, _).items);
```

File: tsconfig.json (hunk 1 above)
Comment (andrewhead) [reply tsconfig.json]: Maybe we don't even need flatMap? Want to check out the suggested fix in data-flow.ts?

File: src/analysis/slice/data-flow.ts (hunk 3 above)
Comment (andrewhead) [reply src/analysis/slice/data-flow.ts]: Also lets you remove the line below (`uses = ...`)

File: tsconfig.json (hunk 1 above)
Comment (joyceerhl) [reply tsconfig.json]: Yep, I removed the extra lib from the tsconfig!

File: src/analysis/slice/data-flow.ts (hunk 3 above)
Comment (joyceerhl) [reply src/analysis/slice/data-flow.ts]: Great point! Dropped the assignment to `usesArr` too.

File: src/test/analysis.test.ts [hunk 4]
```diff
@@ -181,12 +181,20 @@ describe('getDefs', () => {
   describe('detects definitions', () => {
     it('for assignments', () => {
       let defs = getDefsFromStatement('a = 1');
-      expect(defs[0]).to.include({ type: SymbolType.VARIABLE, name: 'a', level: ReferenceType.DEFINITION });
+      expect(defs[0]).to.include({
```
Comment (joyceerhl) [inline src/test/analysis.test.ts]: These are just formatting changes from `jlpm run format:all`

**Comment by joyceerhl** [comment]
> Can you add a unit test for `getUses` to verify the fix on a minimal class?

https://github.com/microsoft/gather/pull/28/files#diff-570ebd4142e7b14c265ba52bd048c158R379-R404


> Can you also run a test to verify that params used in an instance function body aren't registered as "uses" of that param?

I might have misunderstood your request--is this what you had in mind?
https://github.com/microsoft/gather/pull/28/files#diff-570ebd4142e7b14c265ba52bd048c158R419-R426


> Also, if you haven't already, can you verify the regression tests run fine (`jlpm run test`), and prettify the source (`jlpm run format:all`)?

Done and done--thanks for the reminder!

**Pull request merged**